jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java

```diff
@@ -0,0 +1,283 @@
+     * path and add it to {@link #a}. Instead of altering the underlying graph, this implementation
+     * masks edges and nodes that shall not be considered using {@link MaskSubgraph}.
+     */
+    private void tryAddNext()
+    {
+        GraphPath<V, E> lastShortestPath = a.get(a.size() - 1);
+        // The spur node ranges from the first node to the next to last
+        // node in the previous k-shortest path.
+        for (int i = 0; i < lastShortestPath.getLength() - 1; i++) {
+            // Spur node is retrieved from the previous k-shortest path
+            V spurNode = lastShortestPath.getVertexList().get(i);
+            // The sequence of nodes from the source to the spur node of
+            // the previous k-shortest path.
+            List<V> rootPath = lastShortestPath.getVertexList().subList(0, i);
+            List<V> maskedVertices = new ArrayList<V>();
+            List<E> maskedEdges = new ArrayList<E>();
+            for (GraphPath<V, E> path : a) {
+                if (rootPath.equals(path.getVertexList().subList(0, i))) {
+                    // Remove the links that are part of the previous
+                    // shortest paths which share the same root path.
+                    maskedEdges.add(
+                        graph
+                            .getEdge(path.getVertexList().get(i), path.getVertexList().get(i + 1)));
+                }
+            }
+            for (V rootPathNode : rootPath) {
+                if (!rootPathNode.equals(spurNode)) {
+                    maskedVertices.add(rootPathNode);
+                }
+            }
+
+            MaskSubgraph<V, E> maskedSubgraph;
+            if (graph instanceof DirectedGraph) {
+                maskedSubgraph = new DirectedMaskSubgraph<V, E>(
+                    (DirectedGraph<V, E>) graph, maskedVertices::contains, maskedEdges::contains);
+            } else {
+                maskedSubgraph =
+                    new MaskSubgraph<V, E>(graph, maskedVertices::contains, maskedEdges::contains);
+            }
+            // Calculate the spur path from the spur node to the sink.
+            ShortestPathAlgorithm<V, E> maskedShortestSpurPathAlgorithm =
+                spFactory.apply(maskedSubgraph);
+            GraphPath<V, E> spurPath = maskedShortestSpurPathAlgorithm.getPath(spurNode, sink);
+
+            if (null != spurPath && 0 != spurPath.getLength()) { // BellmannFord
+                                                                 // returns
+                                                                 // empty
+                                                                 // path
+                                                                 // instead
+                                                                 // of null
+                // Entire path is made up of the root path and spur path.
+                ArrayList<E> totalPath = new ArrayList<E>();

```

Comment:
This is rather inefficient. You could do the following:
1. you have already calculated the rootPath which is stored as a list of vertices List<V> rootPath.  Consequently you also know the cost of the rootPath C_rootpath
2. You have calculated the spurPath as a GraphPath with cost C_spurPath

You could now simply create a new GraphWalk as follows:
```
List<V> totalPath=new ArrayList<>(rootPath.subList(0,rootPath.size()-1));
totalPath.addAll(spurPath.getVertexList);
GraphWalk<V, E> gw = new GraphWalk<V, E>(graph, source, target, totalPath, C_rootPath+C_spurPath);
```
List<V> totalPath=new ArrayList<>(rootPath.subList(0,rootPath.size()-1));
totalPath.addAll(spurPath.getVertexList);
GraphWalk<V, E> gw = new GraphWalk<V, E>(graph, source, target, totalPath, C_rootPath+C_spurPath);

It is technically quite annoying that the ShortestPath algorithms return a GraphPath instead of a GraphWalk, otherwise you could have simply used the concat function from GraphWalk... We should probably address this.

Reply:
I have tried your suggestion. It is temptingly elegant, but unfortunately not correct. After adding the needed ifs the code looked way uglier than the simple iteration that is in my initial code.
I kept the working code.